fink_filters/filter_anomaly_notification/filter.py

```diff
@@ -0,0 +1,72 @@
+from pyspark.sql.functions import when, lit
+import numpy as np
+from astropy import units as u
+from astropy.coordinates import SkyCoord
+import pandas as pd
+import filter_utils
+
+
+
+def anomaly_notification_(df, threshold=10) -> pd.Series:
+    """
+    Create event notifications with a high anomaly_score value
+    Parameters
+    ----------
+    df : Spark DataFrame with column :
+        objectId : unique identifier for this object
+        lc_features: Dict of dicts of floats.
+            Keys of first dict - filters (fid),
+            keys of inner dicts - names of features
+        rb: RealBogus quality score
+        anomaly_score: Anomaly score
+        timestamp : UTC time
+    threshold : Number of notifications (10 by default)
+
+    Returns
+    ----------
+    out: pandas.Series of bool
+    Return a Pandas DataFrame with the appropriate flag:
+    false for bad alert, and true for good alert.
+    Examples
+    ----------
+    >>> from fink_utils.spark.utils import concat_col
+    >>> from fink_science.ad_features.processor import extract_features_ad
+    >>> from fink_science.anomaly_detection.processor import anomalys_score
+    >>> df = spark.read.format('parquet').load('datatest')
+    >>> what = ['magpsf', 'jd', 'sigmapsf', 'fid']
+    >>> prefix = 'c'
+    >>> what_prefix = [prefix + i for i in what]
+    >>> for colname in what:
+    ...    df = concat_col(df, colname, prefix=prefix)
+    >>> df = df.withColumn('lc_features', extract_features_ad(*what_prefix, 'objectId'))
+    >>> df = df.withColumn("anomaly_score", anomaly_score("lc_features"))
+    >>> df_proc = df.select('objectId', 'candidate.ra',
+    >>>                     'candidate.dec', 'candidate.rb',
+    >>>                     'anomaly_score', 'timestamp')
+    >>> mask = anomaly_notification_(df_proc)
+    >>> print(sum(mask))
+    """
+    med = df.select('anomaly_score').approxQuantile('anomaly_score', [0.5], 0.25)
+    df_filtred = df.sort(['anomaly_score'], ascending=True).limit(min(df.count(), threshold))
+    df_min = df_filtred.toPandas() #Only 10-20 objects fall into the Pandas dataframe
+    filtred_id = set(df_min['objectId'].values)
+    df = df.withColumn('flag',
+                       when((df.objectId.isin(filtred_id)), lit(True))
+                       .otherwise(lit(False)))
```

Comment:
`objectId` is not unique per alert. Several alerts from the same object share the same `objectId`. Hence here you incorporate potentially alerts with low `anomaly_score` because they have the same `objectId` than alerts with high `anomaly_score` (and that's why you probably had more alerts than the `threshold` each time).

The proposed skeleton avoids this.

Reply:
I wanted to output all `objectId` alerts that had at least one alert with a high `anomaly_score`. I thought there shouldn't be many alerts with the same `objectId`, but they could convey useful information.

Reply:
If I use the following approach: 
`pdf_anomalies = df.sort(['anomaly_score'], ascending=True).limit(threshold).toPandas()`
then many of the notifications have the same objectId. So, for example, out of 10 notifications upload `ftransfer_ztf_2023-03-24_521361`, only 4 are associated with a unique `objectId`.
To output only unique objectId, can I use the following code?
```
pdf_dup_group = df_proc.groupBy('objectId').min('anomaly_score')
pdf_dup_group = pdf_dup_group.sort(['min(anomaly_score)'], ascending=True).limit(threshold).toPandas()
upper_bound = np.max(pdf_dup_group['min(anomaly_score)'].values)
df_result = df_proc.withColumn('flag', df_proc['anomaly_score'] <= upper_bound)
pdf_anomalies = df_result.filter(df_result['flag']).toPandas()
```
I want to first group the data by `objectId `and then find the threshold by `anomaly_score`. Then the number of notifications exceeds the threshold, but they will all be high `anomaly_score`. I made this change in the last commit and tested it with `ftransfer_ztf_2023-03-24_521361` — I got 15 notifications at threshold=10

Reply:
Thanks for the detailed explanation -- it looks good to me! The only comment is that you need to scan twice the dataframe, but I think it is unavoidable (I could not come up with another solution).

Reply:
A potential trick -- is to speculate. Say, you know that there shouldn't be more than 10 times more alerts than the threshold, in which case, you can do the following:

```python
# speculate
pdf_anomalies_ext = df.sort(['anomaly_score'], ascending=True).limit(10 * threshold).toPandas()

# Get the upper bound
upper_bound = np.max(pdf_anomalies_ext.drop_duplicates(['objectId'])['anomaly_score'].values[:threshold]) 

# final results
pdf_anomalies = pdf_anomalies_ext[pdf_anomalies_ext['anomaly_score'] <= upper_bound]
```

This way, you scan only once the entire dataframe :-)

Reply:
the speculation simplifies a lot the DAG (the `Exchange` phase -- which brings high pressure on the cluster -- is removed):

```python
df_proc.sort(['anomaly_score'], ascending=True).limit(10 * threshold).explain()                                                                      
== Physical Plan ==
TakeOrderedAndProject(limit=100, orderBy=[anomaly_score#27 ASC NULLS FIRST], output=[candid#0L,schemavsn#1,publisher#2,objectId#3,candidate#4,prv_candidates#5,cutoutScience#6,cutoutTemplate#7,cutoutDifference#8,timestamp#9,cdsxmatch#10,DR3Name#11,Plx#12,e_Plx#13,gcvs#14,vsx#15,x3hsp#16,x4lac#17,mangrove#18,roid#19,rf_snia_vs_nonia#20,snn_snia_vs_nonia#21,snn_sn_vs_all#22,mulens#23,... 11 more fields])
+- FileScan parquet [candid#0L,schemavsn#1,publisher#2,objectId#3,candidate#4,prv_candidates#5,cutoutScience#6,cutoutTemplate#7,cutoutDifference#8,timestamp#9,cdsxmatch#10,DR3Name#11,Plx#12,e_Plx#13,gcvs#14,vsx#15,x3hsp#16,x4lac#17,mangrove#18,roid#19,rf_snia_vs_nonia#20,snn_snia_vs_nonia#21,snn_sn_vs_all#22,mulens#23,... 11 more fields] Batched: false, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[hdfs://134.158.75.222:8020/user/julien.peloton/archive/science/year=2023/month=..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<candid:bigint,schemavsn:string,publisher:string,objectId:string,candidate:struct<jd:double...



df_proc.groupBy('objectId').min('anomaly_score').sort(['min(anomaly_score)'], ascending=True).limit(threshold).explain()                             
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- TakeOrderedAndProject(limit=10, orderBy=[min(anomaly_score)#2589 ASC NULLS FIRST], output=[objectId#3,min(anomaly_score)#2589])
   +- HashAggregate(keys=[objectId#3], functions=[min(anomaly_score#27)])
      +- Exchange hashpartitioning(objectId#3, 200), ENSURE_REQUIREMENTS, [id=#2552]
         +- HashAggregate(keys=[objectId#3], functions=[partial_min(anomaly_score#27)])
            +- Project [objectId#3, anomaly_score#27]
               +- FileScan parquet [objectId#3,anomaly_score#27,month#33,day#34] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[hdfs://134.158.75.222:8020/user/julien.peloton/archive/science/year=2023/month=..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<objectId:string,anomaly_score:double>
```

Reply:
Thank you very much!
That's a great idea, I used this code in the last commit

Reply:
Excellent -- it looks good to me!